gen/field.go

```diff
@@ -94,23 +107,66 @@ func (f fieldGroupGenerator) DefineStruct(g Generator) error {
 			<end>
 		}`,
 		f,
-		TemplateFunc("tag", func(f *compile.FieldSpec) string {
-			// We want to add omitempty if the field is an optional struct or
-			// primitive to redure "null" noise. We won't add omitempty for
-			// optional collections because omitempty doesn't differentiate
-			// between nil and empty collections.
-
-			if (isStructType(f.Type) || isPrimitiveType(f.Type)) && !f.Required {
-				return fmt.Sprintf("`json:\"%s,omitempty\"`", f.Name)
-			}
-
-			return fmt.Sprintf("`json:%q`", f.Name)
-			// TODO(abg): Take go.tag and js.name annotations into account
-		}),
+		TemplateFunc("tag", generateTags),
 		TemplateFunc("declFieldName", f.declFieldName),
 	)
 }
 
+// generateTags parses the annotation on the thrift field and creates the resulting go tag
+func generateTags(f *compile.FieldSpec) (string, error) {
+	ts := []string{}
+
+	jsonTag := compileJSONTag(f, f.Name)
+
+	// process go tags and overwrite json tag if specified in thrift annotation
+	if goAnnotation := f.Annotations[goTagKey]; goAnnotation != "" {
+		tags, err := structtag.Parse(goAnnotation)
+		if err != nil {
+			return "", errors.New(fmt.Sprintf("failed to parse tag %q: %v", goAnnotation, err))
+		}
+		for _, t := range tags.Tags() {
+			if t.Key == jsonTagKey {
+				jsonTag = compileJSONTag(f, t.Name, t.Options...)
+				continue
+			}
+			ts = append(ts, t.String())
+		}
+	}
+
+	ts = append(ts, jsonTag)
+
+	sort.Slice(ts, func(i, j int) bool {
+		return ts[i] < ts[j]
+	})
+
+	// TODO (abg): Take js.name annotation into account
+	return fmt.Sprintf("`%s`", strings.Join(ts, " ")), nil
+}
+
+func compileJSONTag(f *compile.FieldSpec, name string, opts ...string) string {
+	// We want to add omitempty if the field is an optional struct or
+	// primitive to reduce "null" noise. We won't add omitempty for
+	// optional collections because omitempty doesn't differentiate
+	// between nil and empty collections.
+
+	t := structtag.Tag{
+		Key:  jsonTagKey,
+		Name: name,
+		Options: opts,
+	}
+
+	if name == "-" {
+		// If the field name is "-" then it means omit, add no tags
+		return t.String()
+	}
+
+	if (isStructType(f.Type) || isPrimitiveType(f.Type)) && !f.Required {
+		t.Options = append(t.Options, "omitempty")
```

Comment:
We should probably ensure it doesn't already have omitempty with [Tag.HasOption](http://godoc.org/github.com/fatih/structtag#Tag.HasOption).

Plus perhaps a test where the `go.tag` annotation does include `omitempty`.